File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_matlab_labs_and_solutions/adapter/solution/AddressBookAdaptedToTable.m
% AddressBookAdaptedToTable.m

classdef AddressBookAdaptedToTable < Table
	properties (GetAccess = private, SetAccess = private)
		AddressBook % AddressBook
	end

	methods
		% constructor:
		function theAddressBookAdaptedToTable = AddressBookAdaptedToTable(aAddressBook)
			theAddressBookAdaptedToTable.AddressBook = aAddressBook;
		end

		function set.AddressBook(theAddressBookAdaptedToTable, aAddressBook)
			if isempty(aAddressBook)
				error('isempty(aAddressBook)');
			end
			theAddressBookAdaptedToTable.AddressBook = aAddressBook;
		end

		function result = getNumberOfRows(theAddressBookAdaptedToTable)
			result = theAddressBookAdaptedToTable.AddressBook.getNumberOfPersons;
		end

		function result = getNumberOfColumns(theAddressBookAdaptedToTable)
			result = Person.getNumberOfFieldsPerPerson;
		end

		function result = getColumnName(theAddressBookAdaptedToTable, aColumn)
			result = Person.getFieldName(aColumn);
		end

		function result = getValue(theAddressBookAdaptedToTable, aRow, aColumn)
			aPerson = theAddressBookAdaptedToTable.AddressBook.getPerson(aRow);
			result = aPerson.getFieldValue(aColumn);
		end
	end
end

File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/adapter/solution/address_book_adapted_to_table.cpp
// address_book_adapted_to_table.cpp

#include <string>
#include "address_book.h"
#include "address_book_adapted_to_table.h"
#include "dp_exception.h"
#include "person.h"

address_book_adapted_to_table::address_book_adapted_to_table(address_book * a_address_book)
{
	if (a_address_book == nullptr)
	{
		// throw an exception.
	}
	
	my_address_book = a_address_book;
}

int address_book_adapted_to_table::get_number_of_rows() const
{
	return my_address_book->get_number_of_persons();
}

int address_book_adapted_to_table::get_number_of_columns() const
{
	return person::get_number_of_fields_per_person();
}

const std::string & address_book_adapted_to_table::get_column_name(int a_column) const
{
	return person::get_field_name(a_column);
}

const std::string & address_book_adapted_to_table::get_value(int a_row, int a_column) const
{
	person * a_person = my_address_book->get_person(a_row);
	return a_person->get_field_value(a_column);
}